doc/neps/nep-0057-array-in-array-out.rst

```diff
@@ -0,0 +1,231 @@
+for the standard, but explicitly deferred compliance with this aspect.
+
+   We note that one NumPy-specific behavior that remains is returning array
+   scalars rather than 0-D arrays in most cases where the standard, and other
+   array libraries, return 0-D arrays (e.g., indexing and reductions)...
+   There have been multiple discussions over the past year about the
+   feasibility of removing array scalars from NumPy, or at least no longer
+   returning them by default. However, this would be a large effort with some
+   uncertainty about technical risks and impact of the change, and no one has
+   taken it on.
+
+This NEP represents an effort to "take it on". It is a worthwile undertaking:
+scalars "basically duck type 0-D arrays", but they do not *fully* duck type
+zero-rank arrays, with the most fundamental difference being that scalars are
+immutable and zero-rank arrays are not.
+
+It may be argued that if instances of array base class ``np.ndarray`` and
+scalar base class ``np.generic`` were fully interoperable, together, they
+would implement a protocol compatible with the array API standard. Even if this
+were the case, this design is complex and leads to confusion and errors due to
+self-inconsistency (zero-rank array-like scalars are immutable, but arrays of
+other rank are mutable) and inconsistency with all other array API compatible
+libraries. In particular, it leads to difficulties in working with vectorized
+reducing functions, which begin with arrays of rank :math:`N` and return
+objects of rank :math:`M < N`: when :math:`M = 0`, the rules change. This
+prompts an unfortunate pattern of calling ``asarray`` on the results of
+intermediate array operations to ensure that operations like boolean mask
+assignment still work. The inconsistency also presents downstream library
+authors with an unfortunate choice: should they maintain consistency with
+NumPy and prefer to return scalars when possible (e.g. ``scipy.stats``, which
+explicitly uses empty-tuple indexing on all results to *ensure* consistency,
+and ``scipy.special``, which relies on NumPy ufunc machinery), or should they
+follow the lead of the array API standard and prefer zero-rank arrays (e.g.
+``scipy.interpolate``).
+
+Usage and impact
+----------------
+
+Currently, most operations in NumPy involving zero-rank arrays return scalars,
+reducing operations that would naturally result in a zero-rank array actually
+produce a scalar, and indexing operations that would naturally result in a
+zero-rank array actually produce a scalar. The proposal is for these operations
+to return zero-rank arrays instead of scalars.
+
+.. code:: python
+
+   import numpy as np
+   x = np.asarray(1.)
+   np.isscalar(x + x)  # True (main), False (NEP)
+   np.isscalar(np.exp(x))  # True (main), False (NEP)
+   y = np.ones(10)
+   np.isscalar(np.sum(y), axis=-1))  # True (main), False (NEP)
+   np.isscalar(y[0])  # True (main), False (NEP)
+
+For exceptions to these rules, ask Sebastian.
```

Comment:
One exception is `axis=None` -> scalar. 
If I understood correctly, indexing an array like:
```python3
x = np.asarray([1, 2, 3])
x[0]
```
produces a scalar, too?

`axis=None` is one thing, but this would make selling the proposal even harder, since it's a half-measure.

Reply:
> since it's a half-measure.

It makes it much easier because it avoids a swaths of BC issues since `arr1d[0]` will in practice be one of the most common paths to get scalars in user code.
(While it will actually be not that common in major libraries.)

You can call it a half-measure if you like, but I don't see it as such at all.  The point is `x[0, ...]` or `x[..., 0]` is a simple way to get what you want already.

To home in on what IMO, is the core of all of this: **Make scalar returns predictable and avoidable** when working with N-D arrays.
Yes, scalars have more annoyances, but I think focusing on the above makes it easier to see that this is the *minimal* way to achieve it and avoiding the temptation to change things that really don't need changing.

Reply:
> Make scalar returns predictable and avoidable

I don't see how to sell that, though. Individual NumPy operations _are_ pretty predictable right now - operations other than array creation functions will always return a scalar instead of a 0d array. There may be exceptions to that rule, but doesn't that sum it up?

Reply:
You are right, the problem isn't really them being unpredictable, the problem is that there is no good way to avoid the 0-D array -> scalar conversion.
The issue is preserving the 0-D arrays to be able to work with N-D arrays without hassle.

Reply:
> there is no good way to avoid the 0-D array -> scalar conversion.

OK, so do I understand that somewhere in the code, the results of ufuncs and reducing operations *is* a 0d array, and then it gets converted (unavoidably) to a scalar. You are reversing that choice except for reducing operations when `axis` is `None`.

When we fully index an array (e.g. one integer for a 1d array, 2 integers for a 2d array), is the result naturally a 0-d array or a scalar? Is it like ufuncs in that the result is naturally 0d and it gets converted to a scalar, or is it the other way arround - the result is naturally a scalar and if we wanted it to return a 0d array, it would not be so different from just calling `asarray` on the result?